assimilation_code/programs/preprocess/preprocess.f90

```diff
@@ -2,128 +2,188 @@
+!>
+!> The output module for observation types can contain multiple fragments of 
+!> standard F90 that may be required to implement forward observation operators 
+!> for DART. The sections are retrieved from the files by this program and 
+!> inserted into the appropriate blanks in the DEFAULT_obs_def_mod.F90 and 
+!> DEFAULT_obs_kind_mod.F90 templates. 
+!>
 !> The final obs_def_mod.f90 and obs_kind_mod.f90 that are created contain
 !> the default code plus all the code required from the selected observation
-!> type modules. Preprocess also inserts the required identifier and string
-!> for the corresponding observation kinds (and only those kinds).
+!> type modules. Preprocess also inserts the required identifiers and strings
+!> for the corresponding observation quantities.
+!>
+!> @todo FIXME: the module name is set by whatever is in the DEFAULT_xxx file
+!> which is still obs_kind_mod instead of _qty_.  also i have changed the default
+!> filenames here to be _qty_(HK?) but in reality all the pathnames files for the
+!> build system expect _kind_ and the names are always specified in our input.nml
+!> namelist files, so we aren't depending on the defaults. (maybe we should?)
+!> and finally, this code generates module 'use' statements, and those, 
+!> to keep things working, are still outputing "use obs_kind_mod" instead 
+!> of "use obs_qty_mod".   this should change but i'm not sure when.
+!>
 
 program preprocess
 
-! NEED TO ADD IN ALL THE ERROR STUFF
+use types_mod,      only : r8, MISSING_R8     ! @todo FIXME r8 is needed; is MISSING_R8?
+use utilities_mod,  only : register_module, error_handler, E_ERR, E_MSG,   &
+                           file_exist, open_file, close_file,              &
+                           initialize_utilities, do_nml_file, do_nml_term, &
+                           find_namelist_in_file, check_namelist_read,     &
+                           finalize_utilities, log_it
+use parse_args_mod, only : get_args_from_string, get_name_val_pairs_from_string, get_next_arg
 
-use utilities_mod, only : register_module, error_handler, E_ERR, E_MSG,   &
-                          file_exist, open_file, logfileunit,             &
-                          initialize_utilities, finalize_utilities,       &
-                          find_namelist_in_file, check_namelist_read
+implicit none
 
+! Pick something ridiculously large and forget about it (lazy)
+integer, parameter   :: MAX_TYPES = 5000, MAX_QTYS = 5000
 
-implicit none
+! max valid number of tokens per line.  allocate more
+! to handle error conditions.
+integer, parameter   :: MAX_TOKENS = 20
+integer              :: ntokens
+character(len=256)   :: token(MAX_TOKENS)
+character(len=256)   :: valtokens(MAX_TOKENS)
 
-! version controlled file description for error handling, do not edit
-character(len=256), parameter :: source   = &
-   "$URL$"
-character(len=32 ), parameter :: revision = "$Revision$"
-character(len=128), parameter :: revdate  = "$Date$"
```

Comment:
These should not have been removed, they should be fixed.
ALL the error_handler() messages are fundamentally wrong in that the new ones do not mention the offending routine,
they simply report that 'preprocess' is throwing the error. This is inconsistent with the intended use.
Furthermore, without these, the 'register_module()' call requires them, and this makes the program not compile.
I understand adding the register_module() call was done after the testing ...

Reply:
The source, revision, revdate (and associated register_module and error_handler) were removed in commit fe406ec25931d2d4d27e1e51efd7963cd79bbe0a (Sep 14)
from both preprocess.f90 and parse_args.f90.
I'm not sure what the reasoning was. @nancycollins was this just a miscommunication about what svn info should be stripped out?

Reply:
Understood ... however, I am working to rectify the situation ... take no action, please.

Reply:
i see tim is putting these back.  i think this was a very early response to finding out git had no way to expand keywords in the file.  my first reaction was to simply remove them and not call register_module.  tim has been filling in the source filename and leaving the others as "", which is fine.  i usually run with the option that doesn't print the register module strings because i find them unhelpful and cluttering in the log file. but if there is an error, then passing "source" to the error handler seems useful so it can include the source filename which triggered the error.

Reply:
longer term goal - if there is a massive "remove svn keywords" pull request someday, you could add fixing these 3 variables as well. you could leave source as a character string variable with the filename in it, and remove revision and revdate.  this will require removing them from the calls to the error_handler() routines, but i believe since they are optional, the error handler call can just end with   ..., source) and it will still compile.